new/chef-upgrade.md

```diff
@@ -0,0 +1,70 @@
+---
+RFC: unassigned
+Author: Thom May <thom@chef.io>
+Status: Draft
+Type: Standards Track
+---
+
+# Chef should be able to upgrade itself
+
+Users should be able to automatically upgrade to the desired version of chef without having to use third party cookbooks. The upgrade should occur without affecting the normal chef client run, ensuring that a consistent state is preserved.
+
+## Motivation
+
+```
+  As an operator, I want chef versions to be consistent across my infrastructure,
+  so that I can ensure my code works correctly
+```
+```
+  As an operator, I want to easily and safely upgrade chef to my desired version,
+  so that I can easily roll out new versions.
+```
+```
+  As a developer, I want our users to get new features and bug fixes in a timely and safe manner,
+  so that everyone's happy.
+```
+
+## Specification
+
+At the end of each chef run, just before the reboot handler, the chef client should check with an update service to discern whether there is a new version of chef. If there is, the client should decide whether or not to upgrade, based on configuration supplied by the administrator.
```

Comment:
Having the upgrade occur at the beginning of the chef run is key for many users.  Here's the two scenarios I've experienced:

1) System images / AMIs:  You roll out a AMI with Chef 12.1 and you bootstrap outside of knife, as many people do. You want to be able to spin up a new host, have chef upgrade to the version your code requires (12.5 maybe for custom resources), and then have the node converge.  Having the upgrade occur at the end would require you to handle the upgrade outside chef and defeats the purpose of this feature.

2) You want to roll out new chef functionality without first rolling out chef and your environment is not entirely convergent all the time. Again back to the 12.5 custom resource example. You want to start using custom_resources and you're not running 12.5 everywhere.  You shouldn't need to upgrade your fleet first then roll out the custom resources. This becomes really key when you consider that in hardware environments a large number of systems are down due to failures at any time.  If you roll out 12.5 and then bump your cookbooks any node that comes back up will fail.  At a previous job our failure rate hovered from between 2-5% due to node/rack/DC failures.  This was 500 - 1250 nodes at our scale so manually resolving chef run failures wouldn't be an option.